File: app/lib/models/message.dart
/// AetherVerse — 消息数据模型
library;

/// 消息发送者
class MessageSender {
  final String id;
  final String type; // user, agent, system
  final String nickname;
  final String? avatarUrl;
  final bool isAi;
  final String? ownerNickname;

  MessageSender({
    required this.id,
    required this.type,
    required this.nickname,
    this.avatarUrl,
    this.isAi = false,
    this.ownerNickname,
  });

  factory MessageSender.fromJson(Map<String, dynamic> json) {
    return MessageSender(
      id: json['id'] as String,
      type: json['type'] as String? ?? 'user',
      nickname: json['nickname'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
      isAi: json['is_ai'] as bool? ?? false,
      ownerNickname: json['owner_nickname'] as String?,
    );
  }
}

/// 消息
class Message {
  final String id;
  final String? roomId;
  final String? conversationId;
  final MessageSender sender;
  final String msgType; // text, image, system, ai_generated
  final String? content;
  final String? imageUrl;
  final String? replyToId;
  final List<String> mentions;
  final bool isAiGenerated;
  final String? moderationStatus;
  final DateTime createdAt;

  Message({
    required this.id,
    this.roomId,
    this.conversationId,
    required this.sender,
    required this.msgType,
    this.content,
    this.imageUrl,
    this.replyToId,
    this.mentions = const [],
    this.isAiGenerated = false,
    this.moderationStatus,
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as String,
      roomId: json['room_id'] as String?,
      conversationId: json['conversation_id'] as String?,
      sender: MessageSender.fromJson(json['sender'] as Map<String, dynamic>),
      msgType: json['msg_type'] as String? ?? 'text',
      content: json['content'] as String?,
      imageUrl: json['image_url'] as String?,
      replyToId: json['reply_to_id'] as String?,
      mentions: (json['mentions'] as List<dynamic>?)?.cast<String>() ?? [],
      isAiGenerated: json['is_ai_generated'] as bool? ?? false,
      moderationStatus: json['moderation_status'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  bool get isText => msgType == 'text';
  bool get isImage => msgType == 'image';
  bool get isSystem => msgType == 'system';
}

/// 私聊会话
class Conversation {
  final String id;
  final String convType; // user_to_user, user_to_agent
  final ConversationParticipant participant;
  final Message? lastMessage;
  final int unreadCount;
  final DateTime updatedAt;

  Conversation({
    required this.id,
    required this.convType,
    required this.participant,
    this.lastMessage,
    this.unreadCount = 0,
    required this.updatedAt,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      id: json['id'] as String,
      convType: json['conv_type'] as String? ?? 'user_to_user',
      participant: ConversationParticipant.fromJson(
        json['participant'] as Map<String, dynamic>,
      ),
      lastMessage: json['last_message'] != null
          ? Message.fromJson(json['last_message'] as Map<String, dynamic>)
          : null,
      unreadCount: json['unread_count'] as int? ?? 0,
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

/// 会话对方
class ConversationParticipant {
  final String id;
  final String nickname;
  final String? avatarUrl;

  ConversationParticipant({
    required this.id,
    required this.nickname,
    this.avatarUrl,
  });

  factory ConversationParticipant.fromJson(Map<String, dynamic> json) {
    return ConversationParticipant(
      id: json['id'] as String,
      nickname: json['nickname'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
    );
  }
}

File: server/app/models/message.py
"""
SQLAlchemy ORM — 消息 & 聊天系统
2 张表: messages, conversations
"""

import uuid
from datetime import datetime

from sqlalchemy import (
    TIMESTAMP,
    Boolean,
    CheckConstraint,
    Enum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, UUIDMixin
from shared.constants import MessageType, ModerationStatus


class Message(UUIDMixin, Base):
    __tablename__ = "messages"

    room_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    conversation_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    sender_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    sender_type: Mapped[str] = mapped_column(
        Enum("user", "agent", "system", name="sender_type", create_type=False),
        nullable=False,
    )
    msg_type: Mapped[str] = mapped_column(
        Enum(MessageType, name="message_type", create_type=False),
        nullable=False,
        server_default="text",
    )
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    reply_to_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    mentions: Mapped[list | None] = mapped_column(ARRAY(PG_UUID(as_uuid=True)), nullable=True)

    # AIGC 标识
    is_ai_generated: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    aigc_model: Mapped[str | None] = mapped_column(String(64), nullable=True)
    aigc_provider: Mapped[str | None] = mapped_column(String(64), nullable=True)

    # 审核
    moderation_status: Mapped[str | None] = mapped_column(
        Enum(ModerationStatus, name="moderation_status", create_type=False),
        server_default="approved",
    )
    moderation_code: Mapped[str | None] = mapped_column(String(32), nullable=True)

    # 元数据
    client_info: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    request_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=text("NOW()"), nullable=False
    )
    deleted_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)

    __table_args__ = (
        CheckConstraint(
            "(room_id IS NOT NULL AND conversation_id IS NULL) OR "
            "(room_id IS NULL AND conversation_id IS NOT NULL)",
            name="ck_messages_target",
        ),
        Index("idx_messages_room_time", "room_id", "created_at", postgresql_where=text("room_id IS NOT NULL")),
        Index("idx_messages_conv_time", "conversation_id", "created_at",
              postgresql_where=text("conversation_id IS NOT NULL")),
        Index("idx_messages_sender", "sender_id", "sender_type", "created_at"),
        Index("idx_messages_request_id", "request_id", unique=True,
              postgresql_where=text("request_id IS NOT NULL")),
        Index("idx_messages_moderation", "moderation_status",
              postgresql_where=text("moderation_status != 'approved'")),
    )


class Conversation(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "conversations"

    conv_type: Mapped[str] = mapped_column(
        Enum("user_to_user", "user_to_agent", name="conversation_type", create_type=False),
        nullable=False,
    )
    participant_a: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    participant_b: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    participant_b_type: Mapped[str] = mapped_column(
        Enum("user", "agent", "system", name="sender_type", create_type=False),
        nullable=False,
    )
    last_message_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    last_message_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    unread_count_a: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    unread_count_b: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")

    __table_args__ = (
        UniqueConstraint("participant_a", "participant_b", "participant_b_type", name="uq_conversations"),
        Index("idx_conversations_a", "participant_a", "last_message_at"),
        Index("idx_conversations_b", "participant_b", "participant_b_type", "last_message_at"),
    )

